rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,647 @@
+        return Ok((
+            input,
+            MQTTDisconnectData {
+                reason_code: Some(0),
+                properties: None,
+            },
+        ));
+    }
+    match be_u8(input) {
+        Ok((rem, reason_code)) => {
+            // We are checking for 1 because in that case we have a
+            // header plus reason code, but no properties.
+            if remaining_len == 1 {
+                // no properties
+                return Ok((
+                    rem,
+                    MQTTDisconnectData {
+                        reason_code: Some(0),
+                        properties: None,
+                    },
+                ));
+            }
+            match parse_properties(rem, true) {
+                Ok((rem, properties)) => {
+                    return Ok((
+                        rem,
+                        MQTTDisconnectData {
+                            reason_code: Some(reason_code),
+                            properties: properties,
+                        },
+                    ));
+                }
+                Err(e) => return Err(e),
+            }
+        }
+        Err(e) => return Err(e),
+    }
+}
+
+named!(#[inline], pub parse_auth<MQTTAuthData>,
+       do_parse!(
+           reason_code: be_u8
+           >> properties: call!(parse_properties, true)
+           >>  (
+                 MQTTAuthData {
+                   reason_code: reason_code,
+                   properties: properties,
+                 }
+               )
+       ));
+
+pub fn parse_message(input: &[u8], protocol_version: u8, max_msg_size: usize) -> IResult<&[u8], MQTTMessage> {
+    // Parse the fixed header first. This is identical across versions and can
+    // be between 2 and 5 bytes long.
+    match parse_fixed_header(input) {
+        Ok((fullrem, header)) => {
+            let len = header.remaining_length as usize;
+            // This is the length of the fixed header that we need to skip
+            // before returning the remainder. It is the sum of the length
+            // of the flag byte (1) and the length of the message length
+            // varint.
+            let skiplen = 1 + varint_length(len);
+            let message_type = header.message_type;
+            assert_eq!(skiplen, input.len() - fullrem.len());
+
+            // If the remaining length (message length) exceeds the specified
+            // limit, we return a special truncation message type, containing
+            // no parsed metadata but just the skipped length and the message
+            // type.
+            if len > max_msg_size {
+                let msg = MQTTMessage {
+                    header: header,
+                    op: MQTTOperation::TRUNCATED(MQTTTruncatedData {
+                        original_message_type: message_type,
+                        skipped_length: len + skiplen,
+                    }),
+                };
+                return Ok((&input, msg));
```

Comment:
Should not it be `fullrem` instead of `input` ?

Reply:
My intention here was to treat the complete input as if it remained unparsed. This can also be seen by the fact that I am adding the `skiplen` (i.e. the header length) to the parsed value for _remaining length_ which excludes the header.

Reply:
> My intention here was to treat the complete input as if it remained unparsed.

Interesting. Why this choice ?

On my work with HTTP2, I try to parse as much as I can(that is the fixed header)

Reply:
Note we're just looking at the case of `TRUNCATED` and in this case it makes it easier to skip the rest of the input without always having to account for the variable number of 2-5 bytes already parsed for the fixed header. 

Note that we _have_ parsed the header, it's just that we have decided we want to skip it and return the whole size of the input to skip on the request/response handler side. Whether we include the header length in the returned rest or not does not make a difference here IMHO.

Reply:
Maybe I misunderstand something.

Do we still have the `qos` and other header field for this `TRUNCATED` messages ?

Reply:
Yes, we do. These are part of the fixed header (2-5 bytes). So we can always read these, we even have to read them to get the total length of the message -- which is the amount of bytes to be compared against the limit.  In the header there is a field that gives us the _remaining length_ in addition to the header. If the remaining length is less than the limit, we parse the message and return the remainder as always. If it is larger than the limit, we return TRUNCATED with the header values we have and pass the total length of the input to be skipped to the caller (i.e. the `handle_response`/`handle_request`). Hence we from that side we act as if that message was not parsed at all, but still carry the metadata from the fixed header. That is just what I would expect.

Reply:
Ok, thanks for your explanations

Could you please add a simple comment that this is indeed intentional having here `input` instead of `fullrem` ?

Reply:
Absolutely!